### PR TITLE
Add blog post on PTLC payments in exchange for content

### PR DESCRIPTION
**Pull request by yzernik**
Cross-posted from here: https://github.com/squeaknode/squeak/blob/master/docs/trustless-lightning-payments-for-content.md

### CODE REVIEW

**Comment by 0xB10C** [comment]
Cool! Thanks for cross-posting this. I'll review!

File: content/blog/trustless-lightning-payments-for-content.md [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+subtitle: "How PTLC payment channels in Lightning enable decentralized social media."
+date: 2022-01-07T00:00:00+00:00
+authors:
+  - name: Jonathan Zernik
+    github: yzernik
+    twitter: yzernik
+tags:
+  - "Lightning"
+  - "PTLC"
+  - "Bitcoin"
```
Comment (0xB10C) [inline content/blog/trustless-lightning-payments-for-content.md:13]: not sure if the Bitcoin tag is needed on bitcoin-dev blog :)

File: content/blog/trustless-lightning-payments-for-content.md [hunk 2]
```diff
@@ -0,0 +1,357 @@
+---
+
+title: "Trustless Lightning Payments for Content"
+subtitle: "How PTLC payment channels in Lightning enable decentralized social media."
+date: 2022-01-07T00:00:00+00:00
```
Comment (0xB10C) [inline content/blog/trustless-lightning-payments-for-content.md:5]: We might want to move this a bit closer to the actual publishing date once this is ready for merge.

File: content/blog/trustless-lightning-payments-for-content.md [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+With PTLCs, the seller specifies a 32-byte scalar value (`s`) as the "preimage"
+at the creation-time of the invoice. When the buyer decodes the invoice payment
+string, they will obtain the payment point (`p = s*G`). The buyer will only
+obtain the "preimage" `s` after the payment is settled.
+
+So now that we know that PTLC invoices use elliptic curve points, we can
+take advantage of the distributive property of points on elliptic curves:
+
+```
+(s1 + s2)*G = s1*G + s2*G
```
Comment (0xB10C) [inline content/blog/trustless-lightning-payments-for-content.md:105]: I think you should use the LaTeX rendering for this and the other appearances of `s`, `p1`, etc above and below.

See the example blog post or the Schnorr Basic post for example usage. You'd have to enable LaTeX support in the front-matter with `mathjax: true`.

File: content/blog/trustless-lightning-payments-for-content.md [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+#### How squeaks work under the hood
+
+There are two components in a squeak:
+
+* The squeak header
+* The external fields outside the header.
+
+A squeak header has the following fields:
+
+Field Size | Description | Data type | Comments
```
Comment (0xB10C) [inline content/blog/trustless-lightning-payments-for-content.md:180]: I'll see if I can get the tables to render a bit nicer :)

File: content/blog/trustless-lightning-payments-for-content.md [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+A naive approach to accomplish this would be to simply encrypt the content,
+and then create a Lightning invoice with the decryption key as the invoice
+preimage (using HTLCs), and share the invoice with the consumer. Then, when the consumer
+pays the invoice, they would obtain the decryption key, and use it to unlock
+the content.
+
+<!-- In the current implementation of Lightning in LND (using HTLCs), the preimage -->
+<!-- would be specified in the `r_preimage` field of the `AddInvoice` RPC method. -->
+
+{{< center-figure src="/post-data/trustless-lightning-payments/htlc-example-3.png" caption="" height=600 width=600 >}}
```
Comment (0xB10C) [inline content/blog/trustless-lightning-payments-for-content.md:77]:
```suggestion
{{< center-figure src="/post-data/trustless-lightning-payments/htlc-example-3.png" height=600 width=600 >}}
```

or add a caption.

Comment (0xB10C) [reply content/blog/trustless-lightning-payments-for-content.md:77]: Also, seems like there is a lot of transparent padding a the top and bottom. Maybe crop the image to reduce the transparent padding.

image

File: content/blog/trustless-lightning-payments-for-content.md [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+* Bob decodes the payment request string to get the payment point of the invoice, call it `p3`.
+* Bob calculates `s2*G`. If it is equal to `p3 - p1`, then Bob knows that the invoice is valid.
+* Bob pays the Lightning invoice, and gets the value of `s1 + s2` as the preimage.
+* Bob calculates `s1 = (s1 + s2) - s2` to get the decryption key, and decrypts the content.
+
+If Carol tries to give Bob an invalid invoice, Bob will know before he pays the invoice.
+
+The above example assumes that Bob is able to trust that the value of `p1` is valid.
+
+{{< center-figure src="/post-data/trustless-lightning-payments/ptlc-example-2.png" caption="" height=600 width=600 >}}
```
Comment (0xB10C) [inline content/blog/trustless-lightning-payments-for-content.md:129]: Crop transparent padding and (optionally) add a caption.

image

**Review by 0xB10C — COMMENTED** [review]
A few initial comments. Haven't been able to focus on fully reviewing this yet :) 

- I think it's on-topic to publish a LN related post on a bitcoin-dev blog, although, if there were a ln-dev blog, I'd favor publishing it there.
- I think it would be best to export the title image as PNG with a white background and 2:1 dimensions. If you want to further match the other posts, feel free to include a small circular avatar and a black bold title in the image (but you don't have to).
 image

Will follow up with more comments.

**Comment by 0xB10C** [comment]
There is a preview for this PR here https://deploy-preview-10--bitcoin-dev-blog-preview.netlify.app/

File: content/blog/trustless-lightning-payments-for-content.md (hunk 3 above)
Comment (yzernik) [reply content/blog/trustless-lightning-payments-for-content.md]: Done.

File: content/blog/trustless-lightning-payments-for-content.md (hunk 1 above)
Comment (yzernik) [reply content/blog/trustless-lightning-payments-for-content.md]: Done.

File: content/blog/trustless-lightning-payments-for-content.md (hunk 5 above)
Comment (yzernik) [reply content/blog/trustless-lightning-payments-for-content.md]: Fixed.

File: content/blog/trustless-lightning-payments-for-content.md (hunk 6 above)
Comment (yzernik) [reply content/blog/trustless-lightning-payments-for-content.md]: Fixed.

File: content/blog/trustless-lightning-payments-for-content.md (hunk 4 above)
Comment (yzernik) [reply content/blog/trustless-lightning-payments-for-content.md]: I just fixed the table formatting.

File: content/blog/trustless-lightning-payments-for-content.md (hunk 2 above)
Comment (yzernik) [reply content/blog/trustless-lightning-payments-for-content.md:5]: Will do.

**Comment by yzernik** [comment]
@0xB10C Thank you for reviewing! I just resolved the open issues.

I also did a bit more editing to make the blog post flow more smoothly.

**Comment by 0xB10C** [comment]
Thanks! I'll revisit this in the next days and do another round of review before merging.

File: content/blog/trustless-lightning-payments-for-content.md [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+All of these values are used to populate the squeak header. After the header is created:
+
+* The squeak hash is calculated from the header.
+* The private key of the author is used to sign the squeak hash.
+* The signature is then attached to the squeak.
+
+The [MakeSqueak](https://github.com/squeaknode/squeak/blob/fde192b8b59b59090c6d2eb39f87d6b41b15c05e/squeak/core/__init__.py#L415-L462) function looks like this:
+
+```python
+def MakeSqueak(
```
[the remark this answers is not in the document]
Comment (0xB10C) [reply content/blog/trustless-lightning-payments-for-content.md:229]: Like, I understand why you'd want to have the code in documentation like https://github.com/squeaknode/squeak/blob/master/docs/trustless-lightning-payments-for-content.md, but I'm not sure if that's fitting here.

[the remark this answers is not in the document]
Comment (0xB10C) [reply content/blog/trustless-lightning-payments-for-content.md:229]: Also, this code snippet will be outdated as soon as you make a change in your implementation.

File: content/blog/trustless-lightning-payments-for-content.md [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+* Bob already knows $s2$, because the seller sent it to him.
+* Bob calculates $s1 = (s1 + s2) - s2$ to obtain the decryption key.
+* Bob can then decrypt the content of the squeak.
+
+Now Bob can read the content of the squeak!
+
+The [GetDecryptedContent](https://github.com/squeaknode/squeak/blob/fde192b8b59b59090c6d2eb39f87d6b41b15c05e/squeak/core/__init__.py#L242-L261) method of the Squeak class looks like this:
+
+```python
+    def GetDecryptedContent(
```
Comment (0xB10C) [inline content/blog/trustless-lightning-payments-for-content.md:318]: This too is code (implementation details) without a real reference in the text. See comment on `MakeSqueak()` above.

**Review by 0xB10C — CHANGES_REQUESTED** [review]
Did another review round.

Ultimately, the second half of the post is very much a documentation with the protocol field specification and complete functions included. The first part is a bit more educational and could be used as a blog post. However, at this current stage I'd say it is not suitable for bitcoin-dev blog. It needs more work. I'd happy to do another round of review if you want to rework it. Sorry for not catching this in my review 10 days ago - I did think that something was off while skimming this - but couldn't tell what it was.

**Comment by 0xB10C** [comment]
closing this for now as there hasn't been any activity since January

**Comment by yzernik** [comment]
@0xB10C Thanks. I will try to update this PR and reopen in a few weeks when I get some time.

**Pull request closed without merging**